### DISCUSSION TITLE
Fine tuning trained model to new dataset

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#71] [Archived]
<USER>>>> josh_meyer<br/> <DATE> [August 2, 2019, 6:26pm]<br/> <br/> Hey All!<br/> <br/> Has anyone tried fine-tuning a pre-trained Tacotron2 model to new,<br/> smaller dataset? What results are you getting? What *do*'s and *don't*'s<br/> have you discovered?<br/> <br/> This approach is also called adapting the model, speaker adaptation, or<br/> transfer learning.<br/> <br/> I'm currently fine-tuning a pre-trained Tacotron2 to a 15-hour in-house<br/> dataset. After 10K iterations, I can hear the new voice (not the<br/> original LJSpeech), but the model can only produce the first few words<br/> of a sentence. Nevertheless, the amount of progress in just 10K<br/> iterations is promising.<br/> <br/> I'm thinking that learning rate will be very important here... and<br/> possibly 'freezing' certain model layers. Does anyone have experience on<br/> this?<br/> <br/> Best, *slash* <br/> Josh<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/fine-tuning-trained-model-to-new-dataset]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [August 13, 2019, 8:18am]<br/> <br/> how did this experiment end up?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> carlfm01<br/> <DATE> [August 14, 2019, 3:18am]<br/> <br/> Hello, I'm training a Spanish model with 60h (not cleaned), using the<br/> checkpoint from [824c091](https://github.com/mozilla/TTS/tree/824c091)<br/> <br/> Here's the alignment for 280k tacotron2<br/> <br/> Looks like is having a hard time with the silences (from punctuation).<br/> <br/> (909).png.filename}[1399×725 110<br/> <br/> Test audios:<br/> <br/> (908).png.filename}[1543×673 143<br/> <br/> Examples: *slash* <br/> [examples-tux.zip](https://discourse.mozilla.org/uploads/short-url/rtC5U3mPxHs7b2rDmfTXHG7xApA.zip)<br/> (331,8 KB)<br/> <br/> I'll upload the events log soon. Still cleaning the TTS dataset with a<br/> Spanish model of DeepSpeech which is already getting 30h of clean<br/> speech, I'll fine tune and see if I can get more reviewed<br/> transcriptions. *slash* <br/> My plan is to publish it under public domain, any suggestion where to<br/> release it?<br/> <br/> The speaker:<br/> <https://librivox.org/reader/3946?primary_key=3946&search_category=reader&search_page=1&search_form=get_results><br/> <br/> Still need to test WaveRNN, I'll keep updating.<br/> <br/> For all involved, thanks for the amazing work!<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> carlfm01<br/> <DATE> [August 15, 2019, 6:30pm]<br/> <br/> Canceled the training, I see that the silence at the end is the issue<br/> with the punctuation, the trim silence is enabled, insolated the<br/> trimming parts to test it, and yes it is working, the silence is being<br/> removed, any ideas?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> josh_meyer<br/> <DATE> [August 15, 2019, 10:28pm]<br/> <br/> .<br/> <br/> I took the trained (on LJSpeech) Tacotron2 model from<br/> [824c091](https://github.com/mozilla/TTS/tree/824c091), and began<br/> fine-tuning on an in-house *slash* ~15 (male, professionally recorded) dataset.<br/> <br/> At first, I fine-tuned all the weights to the new data with a learning<br/> rate of `0.00001`, and `'lr_decay': false`. This lead to no real<br/> learning and garbage output.<br/> <br/> You can see in the below graph that both training and eval Loss curves<br/> flatline quickly:<br/> <br/> LOSS:<br/> 0.004 +-+-----+--------+-------+-------+-------+--------+-------+-----+-+ <br/> + + + + + + + + + <br/> 0.0035 A-+ eval Loss == A +-+ <br/> | train Loss == B | <br/> | | <br/> 0.003 +-+ +-+ <br/> | | <br/> 0.0025 +-+ +-+ <br/> | | <br/> 0.002 +-+ +-+ <br/> | | <br/> | | <br/> 0.0015 A-+ +-+ <br/> AA | <br/> 0.001 +AA +-+ <br/> BBAAAAAAAAAAAAAA A | <br/> |BBBBBBBBBBBB AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA | <br/> 0.0005 +-+ BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB +-+ <br/> + + + + + + + + + <br/> 0 +-+-----+--------+-------+-------+-------+--------+-------+-----+-+ <br/> 0 50 100 150 200 250 300 350 400<br/> ITERATION #<br/> <br/> Then I tried *only* fine-tuning the `postnet` from the original<br/> pre-trained model, with not great results:<br/> <br/> LOSS:<br/> 0.004 +-+-----+--------+-------+-------+-------+--------+-------+-----+-+ <br/> + + + + + + + + + <br/> | eval Loss A | <br/> 0.0035 +-+ eval Loss B +-+ <br/> | | <br/> 0.003 +-+ +-+ <br/> | | <br/> | | <br/> 0.0025 +-+ +-+ <br/> | | <br/> | | <br/> 0.002 +-+ +-+ <br/> | | <br/> | | <br/> 0.0015 +-A +-+ <br/> |BAA | <br/> 0.001 +-+ AAAAAA +-+ <br/> | B AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA A | <br/> + BBBBBBBBBBBBBBBBB + + + AAA AAAAAAAAAAAA + <br/> 0.0005 +-+-----+--------B--BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB--+-+ <br/> 0 10 20 30 40 50 60 70 80 <br/> ITERATION #<br/> <br/> Then I fine-tuned just the `postnet` and `decoder`:<br/> <br/> LOSS:<br/> 0.012 +A+-----+--------+-------+--------+-------+-------+--------+-----+-+ <br/> + + + + + + + + + <br/> | eval Loss A | <br/> 0.01 +-+ train Loss B +-+ <br/> | | <br/> | | <br/> 0.008 +-+ +-+ <br/> | | <br/> | | <br/> 0.006 +-+ +-+ <br/> | A | <br/> | | <br/> | | <br/> 0.004 +B+ +-+ <br/> | | <br/> | A | <br/> 0.002 +-BAAA +-+ <br/> | BB AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA A | <br/> + BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBAABAAAAAAAAAAAAAAAAA + <br/> 0 +-+-----+--------+-------+--------+-------+-------+--------+-----+-+ <br/> 0 10 20 30 40 50 60 70 80 <br/> ITERATION #<br/> <br/> Finally, I took the above model (i.e. LJSpeech with fine-tuned `postnet`<br/> and `decoder`) and fine-tuned the model above and including the<br/> `attention` layers:<br/> <br/> LOSS:<br/> 0.00095 +-+--------+----------+----------+---------+----------+--------+-+ <br/> A + + + + + + <br/> 0.0009 AAA eval Loss A +-+ <br/> |AAAA train Loss B | <br/> 0.00085 +-+ AAAAA +-+ <br/> | AAAAAAAA | <br/> 0.0008 +-+ AAAAAAA +-+ <br/> | AAAAAAAAA | <br/> 0.00075 +-+ AAAAAAAAAAAAAAA +-+ <br/> | AAAAAAAAAAAAA AAA | <br/> | A AAAAAAAAAAAAAAAAAAAA | <br/> 0.0007 +-+ A A AAAAAAA+-+ <br/> | | <br/> 0.00065 +-+ +-+ <br/> BBBB | <br/> 0.0006 BBBBBBB B +-+ <br/> |B BBBBBBBBBBBBBBBB BBB B BB | <br/> 0.00055 +-+ BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB BBB BBBB BBBBB BB+-+ <br/> + + BBBBB BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB + <br/> 0.0005 +-+--------+----------+--------B-+----B-BB-BBB-BBB--BBBBBBBBBBB+-+ <br/> 0 100 200 300 400 500 600 <br/> ITERATION #<br/> <br/> Here's the same graph above but more detailed and on `tensorboard`:<br/> <br/> This last model was training pretty steadily, but slowly, and with bad<br/> attention alignment:<br/> <br/> 270<br/> <br/> I've found some bad transcripts I'm going to fix (less than 500 out of<br/> 20,000), and then I'm going to fine-tune this most recent checkpoint,<br/> updating the encoder as well.<br/> <br/> My original thought was to not touch the embedding layer or encoder, as<br/> that would be a text-based representation, and I just want to change the<br/> voice, but for the same language (English). However, I read in the<br/> DeepVoice2 paper (funny enough, they made Tacotron2 before Google!) they<br/> injected speaker-specific info in the decoder only, and that performed<br/> badly. They had to add that information to the encoder to get good<br/> multi-speaker results.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [August 16, 2019, 9:45am]<br/> <br/> how do you use Deep Speech to clean TTS dataset? Is it you give the<br/> voice and check the generated transcript?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [August 16, 2019, 9:51am]<br/> <br/> is the dataset multi-speaker?<br/> <br/> I can tell that if you input speaker embedding to encoder, that does not<br/> work perfectly as well. The alignment gets broken for some speakers. I<br/> guess the better way is to add embedding to encoder and decoder<br/> together. I am also working on to enable multi-speaker.<br/> <br/> PS: What is the library printing loss on the terminal ? that looks cool!<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> carlfm01<br/> <DATE> [August 16, 2019, 3:14pm]<br/> <br/> I'm using the Windows Speech recognition to find audio location of the<br/> sentences generated with spacy, then with the obtained<br/> audio/transciption I use DS to mark as valid where the transcription is<br/> the same. Never using related text for the lm, that can bias the<br/> results.<br/> <br/> (911).png.filename}[1467×653 186<br/> <br/> Now only using the cleaned audios using the checkpoint(taco2 260k) looks<br/> better. *slash* <br/> Now the punctuation starts to work properly, I still need to find the<br/> source of the silence at the end. *slash* <br/> [280710.zip](https://discourse.mozilla.org/uploads/short-url/wn0d3MTPWizhX0HoaNkhiD0sZnX.zip)<br/> (471,6 KB)<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> josh_meyer<br/> <DATE> [August 21, 2019, 10:42pm]<br/> <br/> -- the dataset isn't<br/> multi-speaker, no.<br/> <br/> However, it seems there were some mis-aligned pairs of<br/> *slash* <utterance,transcript *slash* > in the data, fixing those has improved things.<br/> <br/> I'm interested in the multi-speaker approach, though. I think it could<br/> be a nice base model to use to adapt to new speakers when you have about<br/> 30 minutes of data. thoughts?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> josh_meyer<br/> <DATE> [August 21, 2019, 10:46pm]<br/> <br/> , can you explain this<br/> issue some more? I see similar padding, but I thought this was normal<br/> for batch training, no?<br/> <br/> , am I missing something?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> carlfm01<br/> <DATE> [August 22, 2019, 1:46am]<br/> <br/> > can you explain this issue some more?<br/> <br/> Hello, I don't know if it is an issue, even with the cleaned set and<br/> silence trimmed I still see the padding.<br/> <br/> > thought this was normal for batch training, no?<br/> <br/> Same question here.<br/> <br/> Sometimes the padding goes away, but as soon as it appears again the<br/> alignment cracks.<br/> <br/> (914).png.filename}[1473×685 169<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [August 22, 2019, 1:13pm]<br/> <br/> it is about batching<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> josh_meyer<br/> <DATE> [August 22, 2019, 6:19pm]<br/> <br/> -- Is there anything in<br/> particular to optimize for padding? any hyperparams? I guess padding<br/> will be less with smaller batches, because the variation within a batch<br/> will be less.<br/> <br/> Batching is done on similar-length audio files, correct? That is, the<br/> examples within a single batch are more similar in length than a random<br/> sample<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [August 22, 2019, 7:27pm]<br/> <br/> there is only `batch_group_size` to play with. As you increase it, you'd<br/> see more padding since it performs shuffle in similar length sentences.<br/> <br/> [Archived Post]<br/>
